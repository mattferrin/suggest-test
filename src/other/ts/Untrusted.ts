/**
 * Derived from kaya3's answer: https://stackoverflow.com/a/63155849
 */
export type Untrusted<T> = T extends Function
  ? T | null | undefined
  : T extends ReadonlyArray<infer U>
  ? UntrustedArray<U> | null | undefined
  : T extends object
  ? UntrustedObject<T> | null | undefined
  : T | null | undefined;
export interface UntrustedArray<T> extends Array<Untrusted<T>> {}
export type UntrustedObject<T> = {
  readonly [P in keyof T]: Untrusted<T[P]> | null | undefined;
};
