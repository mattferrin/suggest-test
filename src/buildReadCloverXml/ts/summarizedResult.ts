import { SummaryItem } from "./undefinedOrSummary";
import { Untrusted } from "./Untrusted";

export type FileArray = ReadonlyArray<{
  readonly $: { readonly path: string };
  readonly metrics: ReadonlyArray<{
    readonly $: {
      readonly statements: number;
      readonly coveredstatements: number;
      readonly conditionals: number;
      readonly coveredConditionals: number;
    };
  }>;
  readonly line: ReadonlyArray<{
    readonly $: { readonly type: "cond" };
  }>;
}>;

export function summarizedResult(
  filteredProjectPackageFile: Untrusted<FileArray>
): ReadonlyArray<SummaryItem | undefined> | undefined {
  return filteredProjectPackageFile?.flatMap((projectPackageFile) => {
    const uncoveredStatements =
      (projectPackageFile?.metrics?.[0]?.$?.statements ?? 0) -
      (projectPackageFile?.metrics?.[0]?.$?.coveredstatements ?? 0);
    const statements = projectPackageFile?.metrics?.[0]?.$?.statements ?? 0;

    const uncoveredConditionals =
      (projectPackageFile?.metrics?.[0]?.$?.conditionals ?? 0) -
      (projectPackageFile?.metrics?.[0]?.$?.coveredConditionals ?? 0);
    const conditionals = projectPackageFile?.metrics?.[0]?.$?.conditionals ?? 0;

    return {
      conditionals,
      path: projectPackageFile?.$?.path,
      statements,
      uncoveredConditionals,
      uncoveredStatements,
      weightedConditionals:
        conditionals === 0
          ? 0
          : Math.pow(uncoveredConditionals, 1.3) / conditionals,
      weightedStatements:
        statements === 0 ? 0 : Math.pow(uncoveredStatements, 1.3) / statements,
    };
  });
}
