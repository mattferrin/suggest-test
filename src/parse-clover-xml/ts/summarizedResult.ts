import { Untrusted } from "../../other/ts/Untrusted";
import { SummaryItem } from "./undefinedOrSummary";

export type FileArray = ReadonlyArray<{
  readonly $: { readonly path: string };
  readonly metrics: ReadonlyArray<{
    readonly $: {
      readonly statements: number;
      readonly coveredstatements: number;
      readonly conditionals: number;
      readonly coveredConditionals: number;
    };
  }>;
  readonly line: ReadonlyArray<{
    readonly $: { readonly type: "cond" };
  }>;
}>;

export function summarizedResult(
  filteredProjectPackageFile: Untrusted<FileArray>
): ReadonlyArray<SummaryItem | undefined> | undefined {
  return filteredProjectPackageFile?.flatMap((projectPackageFile) => {
    return {
      conditionals: projectPackageFile?.metrics?.[0]?.$?.conditionals,
      path: projectPackageFile?.$?.path,
      statements: projectPackageFile?.metrics?.[0]?.$?.statements,
      uncoveredConditionals:
        (projectPackageFile?.metrics?.[0]?.$?.conditionals ?? 0) -
        (projectPackageFile?.metrics?.[0]?.$?.coveredConditionals ?? 0),
      uncoveredStatements:
        (projectPackageFile?.metrics?.[0]?.$?.statements ?? 0) -
        (projectPackageFile?.metrics?.[0]?.$?.coveredstatements ?? 0),
    };
  });
}
