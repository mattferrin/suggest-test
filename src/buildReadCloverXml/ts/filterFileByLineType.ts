import { MeowOutput } from "../../bin/ts/cli";
import { FileArray } from "./summarizedResult";
import { Untrusted } from "./Untrusted";

export function filterFileByLineType(
  input: MeowOutput,
  projectPackage: Untrusted<{
    readonly file: FileArray;
  }>
): Untrusted<FileArray> {
  if (!input.flags.includeStatements) {
    return projectPackage?.file?.filter((projectPackageFile) => {
      return projectPackageFile?.line?.some((line) => {
        return line?.$?.type === "cond";
      });
    });
  } else {
    return projectPackage?.file;
  }
}
