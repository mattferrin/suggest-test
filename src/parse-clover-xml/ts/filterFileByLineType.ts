import { Untrusted } from "../../other/ts/Untrusted";
import { ReadCloverXmlInput } from "./readCloverXml";
import { FileArray } from "./summarizedResult";

export function filterFileByLineType(
  input: ReadCloverXmlInput,
  projectPackage: Untrusted<{
    readonly file: FileArray;
  }>
): Untrusted<FileArray> {
  switch (input.tag) {
    case "existing-balance":
      return projectPackage?.file;
    case "existing-conditionals":
    case "existing-statements":
    case "new-balance":
    case "new-conditionals":
    case "new-statements":
      return projectPackage?.file?.filter((projectPackageFile) => {
        return projectPackageFile?.line?.some((line) => {
          return line?.$?.type === "cond";
        });
      });
  }
}
