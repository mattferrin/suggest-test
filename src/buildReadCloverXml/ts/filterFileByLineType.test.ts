import { filterFileByLineType } from "./filterFileByLineType";
import { FileArray } from "./summarizedResult";
import { Untrusted } from "./Untrusted";

type Input = Parameters<typeof filterFileByLineType>[0];

function snap(input: {
  stabileSortId: string;
  comment: string | null;
  input: Input;
  projectPackage: Parameters<typeof filterFileByLineType>[1];
}): void {
  it(input.stabileSortId, () => {
    /** setup mocks */

    /** unit under test */
    const result = filterFileByLineType(input.input, input.projectPackage);

    /** assertions */
    expect({
      [`${input.comment ?? "TODO: comment"}`]: {
        result,
      },
    }).toMatchSnapshot();
  });
}

snap({
  stabileSortId: "fa22c1bd-307c-4ab7-a396-c312ed3cd567",
  comment:
    "`undefined` result because undefined projectPackage has no file WHEN statements included",
  input: { flags: { includeStatements: true } } as Input,
  projectPackage: undefined,
});
snap({
  stabileSortId: "4162262a-50c4-4f2a-a25a-30ac9880c4c2",
  comment:
    "`undefined` result because null projectPackage has no file WHEN statements included",
  input: { flags: { includeStatements: true } } as Input,
  projectPackage: null,
});
snap({
  stabileSortId: "4598d4d8-11f4-464e-a0c4-f2277d2b6318",
  comment:
    "`projectPackage arg file prop` result because file defined WHEN statements included",
  input: { flags: { includeStatements: true } } as Input,
  projectPackage: { file: "projectPackage arg file prop" as any },
});

snap({
  stabileSortId: "7485ae3a-2133-470f-bb52-afd80b6167d0",
  comment:
    "`undefined` result because undefined projectPackage WHEN statements excluded",
  input: { flags: { includeStatements: false } } as Input,
  projectPackage: undefined,
});
snap({
  stabileSortId: "2cd2e1e1-00e4-4ec8-a255-73aa1c778846",
  comment:
    "`undefined` result because null projectPackage WHEN statements excluded",
  input: { flags: { includeStatements: false } } as Input,
  projectPackage: null,
});

snap({
  stabileSortId: "97775dd4-7c63-45e2-904d-f1988ee098c7",
  comment:
    "`undefined` result because undefined projectPackage file WHEN statements excluded",
  input: { flags: { includeStatements: false } } as Input,
  projectPackage: { file: undefined },
});
snap({
  stabileSortId: "044bbd18-d569-4d8f-98af-312f20fe84dd",
  comment:
    "`undefined` result because null projectPackage file WHEN statements excluded",
  input: { flags: { includeStatements: false } } as Input,
  projectPackage: { file: null },
});

snap({
  stabileSortId: "0b619f9c-cccf-4d54-a6ac-00d0f92d8d13",
  comment: "`[]` result because file empty AND statements excluded",
  input: { flags: { includeStatements: false } } as Input,
  projectPackage: { file: [] },
});

snap({
  stabileSortId: "6cc23ee9-8cb5-4bbe-ad01-6153a2fbe6be",
  comment: "`[]` result WHEN item undefined AND statements included",
  input: { flags: { includeStatements: false } } as Input,
  projectPackage: { file: [undefined] },
});
snap({
  stabileSortId: "9d1c1d27-94a7-4b79-9108-93af2e34d88f",
  comment: "`[]` result WHEN item null AND statements included",
  input: { flags: { includeStatements: false } } as Input,
  projectPackage: { file: [null] },
});

snap({
  stabileSortId: "e2152348-6515-4ebb-b1f9-f7aec06b58f9",
  comment: "`[]` result WHEN item line undefined AND statements included",
  input: { flags: { includeStatements: false } } as Input,
  projectPackage: { file: [{ line: undefined }] as Untrusted<FileArray> },
});
snap({
  stabileSortId: "7951520c-b8d6-4015-892b-73186c449e1c",
  comment: "`[]` result WHEN item line null AND statements included",
  input: { flags: { includeStatements: false } } as Input,
  projectPackage: { file: [{ line: null }] as Untrusted<FileArray> },
});

snap({
  stabileSortId: "d7c4b5f0-d1ae-4443-986c-83d7dcaf82ee",
  comment: "`[]` result WHEN item line item undefined AND statements included",
  input: { flags: { includeStatements: false } } as Input,
  projectPackage: { file: [{ line: [undefined] }] as Untrusted<FileArray> },
});
snap({
  stabileSortId: "01b5eabe-78e0-4dec-8b60-915f5c5fbef6",
  comment: "`[]` result WHEN item line item null AND statements included",
  input: { flags: { includeStatements: false } } as Input,
  projectPackage: { file: [{ line: [null] }] as Untrusted<FileArray> },
});

snap({
  stabileSortId: "cd44ae3c-07f7-4a18-a1d2-3bbec8c78fa7",
  comment:
    "`[]` result WHEN item line item $ undefined AND statements included",
  input: { flags: { includeStatements: false } } as Input,
  projectPackage: {
    file: [{ line: [{ $: undefined }] }] as Untrusted<FileArray>,
  },
});
snap({
  stabileSortId: "c700bcf2-e1ff-461d-ae79-e404515dcfc6",
  comment: "`[]` result WHEN item line item $ null AND statements included",
  input: { flags: { includeStatements: false } } as Input,
  projectPackage: { file: [{ line: [{ $: null }] }] as Untrusted<FileArray> },
});

snap({
  stabileSortId: "2e2f7fc5-15ed-438e-abe6-2acfcff9815d",
  comment:
    "`[]` result WHEN item line item $ type IS NOT cond AND statements included",
  input: { flags: { includeStatements: false } } as Input,
  projectPackage: {
    file: [
      { line: [{ $: { type: "NOT cond" as any } }] },
    ] as Untrusted<FileArray>,
  },
});
snap({
  stabileSortId: "e16ee10b-f2fe-4ca2-b968-0f093a669cd9",
  comment:
    "result includes conditionals WHEN item line item $ type IS cond AND statements included",
  input: { flags: { includeStatements: false } } as Input,
  projectPackage: {
    file: [{ line: [{ $: { type: "cond" } }] }] as Untrusted<FileArray>,
  },
});
