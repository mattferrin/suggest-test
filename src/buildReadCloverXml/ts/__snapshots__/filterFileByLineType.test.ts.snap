// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`01b5eabe-78e0-4dec-8b60-915f5c5fbef6 1`] = `
{
  "\`[]\` result WHEN item line item null AND statements included": {
    "result": [],
  },
}
`;

exports[`044bbd18-d569-4d8f-98af-312f20fe84dd 1`] = `
{
  "\`undefined\` result because null projectPackage file WHEN statements excluded": {
    "result": undefined,
  },
}
`;

exports[`0b619f9c-cccf-4d54-a6ac-00d0f92d8d13 1`] = `
{
  "\`[]\` result because file empty AND statements excluded": {
    "result": [],
  },
}
`;

exports[`2cd2e1e1-00e4-4ec8-a255-73aa1c778846 1`] = `
{
  "\`undefined\` result because null projectPackage WHEN statements excluded": {
    "result": undefined,
  },
}
`;

exports[`2e2f7fc5-15ed-438e-abe6-2acfcff9815d 1`] = `
{
  "\`[]\` result WHEN item line item $ type IS NOT cond AND statements included": {
    "result": [],
  },
}
`;

exports[`6cc23ee9-8cb5-4bbe-ad01-6153a2fbe6be 1`] = `
{
  "\`[]\` result WHEN item undefined AND statements included": {
    "result": [],
  },
}
`;

exports[`9d1c1d27-94a7-4b79-9108-93af2e34d88f 1`] = `
{
  "\`[]\` result WHEN item null AND statements included": {
    "result": [],
  },
}
`;

exports[`4598d4d8-11f4-464e-a0c4-f2277d2b6318 1`] = `
{
  "\`projectPackage arg file prop\` result because file defined WHEN statements included": {
    "result": "projectPackage arg file prop",
  },
}
`;

exports[`7485ae3a-2133-470f-bb52-afd80b6167d0 1`] = `
{
  "\`undefined\` result because undefined projectPackage WHEN statements excluded": {
    "result": undefined,
  },
}
`;

exports[`97775dd4-7c63-45e2-904d-f1988ee098c7 1`] = `
{
  "\`undefined\` result because undefined projectPackage file WHEN statements excluded": {
    "result": undefined,
  },
}
`;

exports[`4162262a-50c4-4f2a-a25a-30ac9880c4c2 1`] = `
{
  "\`undefined\` result because null projectPackage has no file WHEN statements included": {
    "result": undefined,
  },
}
`;

exports[`7951520c-b8d6-4015-892b-73186c449e1c 1`] = `
{
  "\`[]\` result WHEN item line null AND statements included": {
    "result": [],
  },
}
`;

exports[`c700bcf2-e1ff-461d-ae79-e404515dcfc6 1`] = `
{
  "\`[]\` result WHEN item line item $ null AND statements included": {
    "result": [],
  },
}
`;

exports[`cd44ae3c-07f7-4a18-a1d2-3bbec8c78fa7 1`] = `
{
  "\`[]\` result WHEN item line item $ undefined AND statements included": {
    "result": [],
  },
}
`;

exports[`d7c4b5f0-d1ae-4443-986c-83d7dcaf82ee 1`] = `
{
  "\`[]\` result WHEN item line item undefined AND statements included": {
    "result": [],
  },
}
`;

exports[`e16ee10b-f2fe-4ca2-b968-0f093a669cd9 1`] = `
{
  "result includes conditionals WHEN item line item $ type IS cond AND statements included": {
    "result": [
      {
        "line": [
          {
            "$": {
              "type": "cond",
            },
          },
        ],
      },
    ],
  },
}
`;

exports[`e2152348-6515-4ebb-b1f9-f7aec06b58f9 1`] = `
{
  "\`[]\` result WHEN item line undefined AND statements included": {
    "result": [],
  },
}
`;

exports[`fa22c1bd-307c-4ab7-a396-c312ed3cd567 1`] = `
{
  "\`undefined\` result because undefined projectPackage has no file WHEN statements included": {
    "result": undefined,
  },
}
`;
